openwisp_notifications/base/models.py

```diff
@@ -105,36 +105,74 @@ def target_url(self):
     def message(self):
         return self.get_message()
 
+    @cached_property
+    def rendered_description(self):
+        if not self.description:
+            return
+        md_description = self._get_formatted_string(self.description)
+        return mark_safe(markdown(md_description))
+
     @property
     def email_message(self):
         return self.get_message(email_message=True)
 
+    def _set_extra_attributes(self, email_message=False):
+        """
+        This method sets extra attributes for the notification
+        which are required to render the notification correctly.
+        """
+        self.actor_link = self.actor_url
+        self.action_link = self.action_url
+        self.target_link = (
+            self.target_url if not email_message else self.redirect_view_url
+        )
+
+    def _delete_extra_attributes(self):
+        """
+        This method removes the extra_attributes set by
+        _set_extra_attributes
+        """
+        for attr in ('actor_link', 'action_link', 'target_link'):
+            delattr(self, attr)
+
+    def _get_formatted_string(self, string, email_message=False):
+        self._set_extra_attributes(email_message=email_message)
+        data = self.data or {}
+        try:
+            formatted_output = string.format(notification=self, **data)
+        except AttributeError as exception:
+            self._invalid_notification(
+                self.pk,
+                exception,
+                'Error encountered in rendering notification message',
+            )
+        self._delete_extra_attributes()
+        return formatted_output
```

Comment:
This is required because we handle the target URL differently when we send the email. 

The email includes a ridrect URL, which flags the notification as read and then redirects the user to the target object.